### PR TITLE
[SYCL] Align nan builtin with the updated specification

### PR DESCRIPTION
**Pull request by aelovikov-intel**
Specification has been changed in
https://github.com/KhronosGroup/SYCL-Docs/pull/519.

### CODE REVIEW

**Comment by aelovikov-intel** [comment]
@steffenlarsen , this is ready for review.

File: sycl/include/sycl/builtins_legacy_scalar.hpp [hunk 1]
```diff
@@ -322,8 +322,30 @@ modf(T x, T2 iptr) {
...
+std::enable_if_t<detail::is_nan_type_v<T> &&
+                     detail::is_non_deprecated_nan_type_v<T>,
+                 detail::nan_return_t<T>>
+nan(T nancode) {
+  return __sycl_std::__invoke_nan<detail::nan_return_t<T>>(
+      detail::convert_data_type<T, detail::nan_argument_base_t<T>>()(nancode));
+}
+template <typename T>
+__SYCL_DEPRECATED(
+    "This is a deprecated argument type for SYCL nan built-in function.")
```
Comment (steffenlarsen) [inline sycl/include/sycl/builtins_legacy_scalar.hpp:345]: This deprecation might be confusion. With overloads we can rely on implicit conversion, while here the users might be told their code is deprecated while their call wouldn't be deprecated when the preview changes are promoted.

File: sycl/test-e2e/Basic/built-ins/marray_math.cpp [hunk 2]
```diff
@@ -136,8 +136,13 @@ int main() {
   sycl::marray<int, 3> ma4{1, 1, 1};
   sycl::marray<float, 3> ma5{180, -180, -180};
   sycl::marray<float, 3> ma6{1.4f, 4.2f, 5.3f};
+#ifndef __INTEL_PREVIEW_BREAKING_CHANGES
   sycl::marray<unsigned int, 3> ma7{1, 2, 3};
   sycl::marray<unsigned long int, 3> ma8{1, 2, 3};
+#else
+  sycl::marray<uint32_t, 3> ma7{1, 2, 3};
+  sycl::marray<uint64_t, 3> ma8{1, 2, 3};
+#endif
```
Comment (steffenlarsen) [inline sycl/test-e2e/Basic/built-ins/marray_math.cpp:145]: Do we need this branching? Seems like we make the changes for both preview and main path. We could keep the old marrays for non-preview testing, despite deprecation.

Same question for the other test case.

File: sycl/include/sycl/builtins_legacy_scalar.hpp (hunk 1 above)
Comment (steffenlarsen) [reply sycl/include/sycl/builtins_legacy_scalar.hpp:345]: Seems to only affect `unsigned long long` which would cause ambiguity error anyway.

**Comment by coldav** [comment]
Hi @aelovikov-intel, I'm not sure if you are aware yet, but this change seems to break SYCL-CTS in math_tests_api (math_builtin_float_double.cpp):
```
build_sycl_cts/tests/math_builtin_api/math_builtin_float_double.cpp:18869:47: error: no matching function for call to 'nan'
 18869 |         static_assert(std::is_same_v<decltype(sycl::nan(inputData_0)), sycl::marray<double, 2>>,
```

**Pull request merged**